src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/ConsiderPassingBaseTypesAsParameters.cs

```diff
@@ -0,0 +1,225 @@
+                    {
+                        typePool.Add(typeSymbol);
+                    }
+                }
+
+                void AddDeclaringTypeSymbol(ISymbol symbol)
+                    => AddType(FindDeclaringType(symbol));
+            }, OperationKind.ParameterReference);
+
+            context.RegisterOperationBlockEndAction(context =>
+            {
+                try
+                {
+                    var overloads = methodSymbol.GetOverloads();
+
+                    foreach (var (parameter, usedTypesPool) in typesUsedPerParameter)
+                    {
+                        var usedTypes = usedTypesPool.ToImmutableArray();
+                        // If type is unused, we ignore it. There is another rule to detect
+                        // unused parameters.
+
+                        if (usedTypes.Length == 1)
+                        {
+                            ReportIfTypesAreDifferent(parameter, usedTypes[0], overloads);
+                        }
+                        else if (usedTypes.Length > 1)
+                        {
+                            ReportIfTypesAreDifferent(parameter, FindMostGenericType(parameter.Type, usedTypes), overloads);
+                        }
+                    }
+                }
+                finally
+                {
+                    typesUsedPerParameter.Free(context.CancellationToken);
+                }
+
+                // Local functions
+                void ReportIfTypesAreDifferent(IParameterSymbol parameter, ITypeSymbol suggestedBaseTypeSymbol, IEnumerable<IMethodSymbol> overloads)
+                {
+                    if (SymbolEqualityComparer.Default.Equals(suggestedBaseTypeSymbol, parameter.Type))
+                    {
+                        return;
+                    }
+
+                    if (suggestedBaseTypeSymbol is INamedTypeSymbol suggestedNamedType)
+                    {
+                        var parameterIndex = methodSymbol.GetParameterIndex(parameter);
+                        var matchingOverload = methodSymbol.GetMatchingOverload(overloads, parameterIndex, suggestedNamedType, context.CancellationToken);
+
+                        if (matchingOverload != null)
+                        {
+                            return;
+                        }
+                    }
+
+                    // All calls on this parameter are using the same base type which is different
+                    // from the one declared, so we can simply report.
+                    context.ReportDiagnostic(parameter.CreateDiagnostic(Rule, parameter.Name,
+                        parameter.Type.Name, suggestedBaseTypeSymbol.Name));
+                }
+            });
+        }
+
+        private static ITypeSymbol FindDeclaringType(ISymbol symbol)
+        {
+            if (symbol.IsOverride)
+            {
+                return FindDeclaringType(symbol.GetOverriddenMember());
```

Comment:
I believe there are some error cases where `GetOverriddenMember` can return null even if `IsOverride` is true. I think you may want to add a null check somewhere.

Reply:
The various properties followed by `GetOverriddenMember`  are not annotated with null, if you know it can return null maybe our helper should be annotated so we avoid such mistake.

Reply:
Ah nullable are not fully enable on roslyn...